### ISSUE TITLE
New idiom for detection of being on device

### ISSUE THREAD

**Issue by DavidPoliakoff** [kokkos/kokkos#3404]
As mentioned [here](https://github.com/kokkos/kokkos/pull/3400#issuecomment-696202053) we're working on a good way to detect whether code is running on the device. This is being iterated on.

I'm filing this issue so that we can have the conversation in public, what do we want that idiom to look like? It seems @crtrott has a design he likes that should span CUDA/DPC++/HIP/OpenMPTarget.

We need this implementation to enable other features

**Comment by DavidPoliakoff** [comment]
It seems @crtrott does not in fact have a solution... :p

**Comment by crtrott** [comment]
So one idea was to have some kind of macro like this:
```
KOKKOS_IF_ON_HOST
{
}
KOKKOS_ELSE_ON_DEVICE
{
}
```

**Comment by crtrott** [comment]
Generally we need to combine these two paradigms:
```
#if !defined( __CUDA_ARCH__)
...
#else
...
#endif 
```
and
```
if constexpr (omp_is_initial_device()) {
...
} else {
...
}
```

**Comment by crtrott** [comment]
Something really ugly would be:
```
KOKKOS_IF_ON_HOST_CONSTEXPR
#if defined(KOKKOS_IF_ON_HOST_MACRO)
{
}
#endif 
```
and 

```
KOKKOS_IF_ON_DEVICE_CONSTEXPR
#if defined(KOKKOS_IF_ON_DEVICE_MACRO)
{
}
#endif
```

**Comment by crtrott** [comment]
Actually what about:
```c++
#if defined(__CUDA_ARCH__) || defined(..)
#define KOKKOS_IS_ON_HOST(A)
#define KOKKOS_IS_NOT_ON_HOST(A) {A}
#endif

#ifdef KOKKOS_ENABLE_OPENMPTARGET
#define KOKKOS_IS_ON_HOST(A) if constexpr (omp_is_initial_dvice()) {A}
#define KOKKOS_IS_NOT_ON_HOST(A) if constexpr (!omp_is_initial_dvice()) {A}
#endif

#ifndef KOKKOS_IS_ON_HOST
#define KOKKOS_IS_ON_HOST(A) {A}
#define KOKKOS_IS_NOT_ON_HOST(A)
#endif
```

Usage:
```c++
  KOKKOS_IS_ON_HOST(
     // do host only stuff probably no use of macros though
  )
  KOKKOS_IS_NOT_ON_HOST(
    // do device only stuff
  )
```

**Comment by DavidPoliakoff** [comment]
Would that have the effect of making all error messages and the like report
on the line of the macro, rather than the line of the code?

**Comment by crtrott** [comment]
yeah. but this is not really meant for users.

**Comment by crtrott** [comment]
and you can call an inline function, so the actual error appears in the body of the function

**Comment by crtrott** [comment]
This seems to work. Note the protected code can't use #ifdefs itself, but it could call functions which use ifdefs internally. 
```c++
#include <Kokkos_Core.hpp>
#include <cmath>

#if defined(__CUDA_ARCH__)
#define KOKKOS_IS_ON_HOST(A)
#define KOKKOS_IS_NOT_ON_HOST(A) {A}
#endif

#ifdef KOKKOS_ENABLE_OPENMPTARGET
#define KOKKOS_IS_ON_HOST(A) if constexpr (omp_is_initial_device()) {A}
#define KOKKOS_IS_NOT_ON_HOST(A) if constexpr (!omp_is_initial_device()) {A}
#endif

#ifndef KOKKOS_IS_ON_HOST
#define KOKKOS_IS_ON_HOST(A) {A}
#define KOKKOS_IS_NOT_ON_HOST(A)
#endif

KOKKOS_INLINE_FUNCTION
void atom_add(int* ptr, int val) {
  KOKKOS_IS_ON_HOST(
    __atomic_fetch_add(ptr,val,__ATOMIC_RELAXED);
    __atomic_fetch_add(ptr,val,__ATOMIC_RELAXED);
  )
  KOKKOS_IS_NOT_ON_HOST(
    Kokkos::atomic_add(ptr,val);
    Kokkos::atomic_add(ptr,val);
  )
}
template<class ExecSpace>
void run_test(ExecSpace space, int N) {
    Kokkos::View<int,ExecSpace> a("A");
    Kokkos::parallel_for(Kokkos::RangePolicy<ExecSpace>(0,N),KOKKOS_LAMBDA(int) {
      atom_add(a.data(),3);
    });
    int result;
    Kokkos::deep_copy(result,a);
    printf("%s %i %i\n",space.name(),result,N*3*2);
}

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc, argv);
  {
    int N = argc > 1 ? atoi(argv[1]) : 1000;
    int R = argc > 2 ? atoi(argv[2]) : 10;
    run_test(Kokkos::DefaultExecutionSpace(),N);
    run_test(Kokkos::DefaultHostExecutionSpace(),N);
  }
  Kokkos::finalize();
}
```

**Comment by mhoemmen** [comment]
Please do the `do { A } while(false)` or comparable trick, so macro invocations can end with a semicolon (and not anger IDEs too much) : - )

**Comment by DavidPoliakoff** [comment]
```
yeah. but this is not really meant for users.
```

Aww, that's _adorable_ :) . It's nice, that after this much time developing software, we still have the capacity to hope.

Anyway, I'm lightly in favor of _requiring_ this thing to call a function (`KOKKOS_IS_ON_HOST_CALL(func,args...)`). Several people are going to write

```c++
KOKKOS_IS_ON_HOST(
  // 1500 lines
);
```

And get confusing error messages, and file the following issue

```
So I was using the functionality in Kokkos to change execution paths based on whether I was on the host or device, and now all of my error messages refer to the first line of this massive region. Wat do?
```

And we'll tell them to use an inline function. I'd rather just force the function call. I know it's clunky, we may just want to put the onus on users to do the right thing (and just answer those issues).

**Comment by crtrott** [comment]
@mhoemmen It would force the macro to end with a semicolon though, otherwise its missing one right?
```
KOKKOS_IS_ON_HOST(
  bar();
)
foo();
```

Also: you are not really supposed to use it anyway ;-)

**Comment by DavidPoliakoff** [comment]
If we're saying "don't use this," we should `IMPL_` prefix it. If we IMPL_ prefix it I don't really care whether we force it to be a function boundary

**Comment by mhoemmen** [comment]
@crtrott wrote:
>  It would force the macro to end with a semicolon though,

No -- the idea is that invocations of the macro need to end with a semicolon.  Here is a brief example:
```cpp
#define MY_MACRO( A ) do { \
  std::cout << "START" << std::endl; \
  A; \
  std::cout << "END" << std::endl; \
} while (false)
```
and you use it like this:
```cpp
MY_MACRO( foo() );
```

**Comment by mhoemmen** [comment]
> Aww, that's _adorable_ :) . It's nice, that after this much time developing software, we still have the capacity to hope.

Maybe y'all remember the example of the STK function called something like `this_function_is_only_for_jane_doe_if_you_are_not_named_jane_do_not_call`.  STK team got a help ticket from someone not named Jane, complaining that the function didn't behave as expected.

**Comment by stanmoore1** [comment]
Is there a reason why users can't have this? See https://github.com/lammps/lammps/blob/master/src/KOKKOS/kokkos_type.h#L1152, would be nice to get rid of this code.

**Comment by DavidPoliakoff** [comment]
> > Aww, that's _adorable_ :) . It's nice, that after this much time developing software, we still have the capacity to hope.
> 
> Maybe y'all remember the example of the STK function called something like `this_function_is_only_for_jane_doe_if_you_are_not_named_jane_do_not_call`. STK team got a help ticket from someone not named Jane, complaining that they tried calling the function but it didn't behave as expected.

The least they could have done was change their name to Jane, and checked if the error sill happened smh my head

**Comment by nliber** [comment]
The only knob we have in SYCL is the macro `__SYCL_DEVICE_ONLY__`.  Unfortunately, the opposite of that is code that is on both the Host AND Device.

**Comment by nliber** [comment]
I ran the following test program:

```
#include <CL/sycl.hpp>

int main() {
    sycl::queue q;

    auto* ucp = static_cast<unsigned char*>(
        sycl::malloc(1, q, sycl::usm::alloc::shared));
    *ucp = 1 << 0;

    q.submit([&](sycl::handler& cgh) {
        cgh.single_task([=] {
            *ucp |= 1 << 1;
#ifdef __SYCL_DEVICE_ONLY__
            *ucp |= 1 << 2;
#endif  // __SYCL_DEVICE_ONLY__
#ifndef __SYCL_DEVICE_ONLY__
            *ucp |= 1 << 3;
#endif  // !__SYCL_DEVICE_ONLY__
        });
    });
    q.wait();

    std::cout << +*ucp << '\n';

    sycl::free(ucp, q);

    return 0;
}
```

and got a result of `7`.

`__SYCL_DEVICE_ONLY__` means Device only
!__SYCL_DEVICE_ONLY__` means Host only

**Comment by trws** [comment]
That seems, interesting.  Is it undefined if the code is host & device both?

For OpenMP, the if constexpr isn't guaranteed to compile, we didn't define `is_initial_device()` as constexpr.  When 5.1 is supported, the mechanism that will be supported would be to bound the code with `begin/end declare variant` directives, which will elide the code if the match conditions don't match.

Also, for cuda, it's worth noting that cub recently refactored their version of this to use this macro https://github.com/NVIDIA/cub/blob/main/cub/util_arch.cuh#L68. It also uses a slightly different pattern with regular non-constant if statements.

**Comment by PhilMiller** [comment]
Fix released in 3.6
